File: MongoBooks2/AngularMongoBooks3/Controllers/DataClasses/DeltaBooks.cs
namespace AngularMongoBooks3.Controllers.DataClasses
{
    using System;

    using BooksCore.Books;

    public class CategoryTotal
    {
        public string Name { get; set; }
        public int TotalPages { get; set; }
        public int TotalBooks { get; set; }
        public float PercentagePages { get; set; }
        public float PercentageBooks { get; set; }

        public CategoryTotal()
        {
            Name = string.Empty;
            TotalPages = 0;
            TotalBooks = 0;
            PercentagePages = 0f;
            PercentageBooks = 0f;
        }

        public CategoryTotal(CategoryTotal src)
        {
            Name = src.Name;
            TotalPages = src.TotalPages;
            TotalBooks = src.TotalBooks;
            PercentagePages = src.PercentagePages;
            PercentageBooks = src.PercentageBooks;
        }
    }

    public class DeltaBooks
    {
        public DateTime Date { get; set; }

        public DateTime StartDate { get; set; }

        public int DaysSinceStart { get; set; }

        public TallyDelta OverallTally { get; set; }

        public TallyDelta LastTenTally { get; set; }

        public CategoryTotal[] LanguageTotals { get; set; }

        public CategoryTotal[] CountryTotals { get; set; }

        public DeltaBooks()
        {
            Date = DateTime.Now;
            StartDate = DateTime.Now;
            DaysSinceStart = 0;
            OverallTally = new TallyDelta();
            LastTenTally = new TallyDelta();
            LanguageTotals = new CategoryTotal[] { };
            CountryTotals = new CategoryTotal[] { };
        }

        public DeltaBooks(BooksDelta booksDelta)
        {
            Date = booksDelta.Date;
            StartDate = booksDelta.StartDate;
            DaysSinceStart = booksDelta.DaysSinceStart;
            OverallTally = new TallyDelta(booksDelta.OverallTally);
            LastTenTally = new TallyDelta(booksDelta.LastTenTally);

            int languageTotalsCount = booksDelta.OverallTally.LanguageTotals.Count;
            LanguageTotals = new CategoryTotal[languageTotalsCount];
            for (int i = 0; i < languageTotalsCount; i++)
            {
                var languageTotal = booksDelta.OverallTally.LanguageTotals[i];
                LanguageTotals[i] =
                    new CategoryTotal
                    {
                        Name = languageTotal.Item1,
                        TotalBooks = (int)languageTotal.Item2,
                        PercentageBooks = (float)languageTotal.Item3,
                        TotalPages = (int)languageTotal.Item4,
                        PercentagePages = (float)languageTotal.Item5
                    };
            }


            int countryTotalsCount = booksDelta.OverallTally.CountryTotals.Count;
            CountryTotals = new CategoryTotal[countryTotalsCount];
            for (int i = 0; i < countryTotalsCount; i++)
            {
                var countryTotal = booksDelta.OverallTally.CountryTotals[i];
                CountryTotals[i] =
                    new CategoryTotal
                    {
                        Name = countryTotal.Item1,
                        TotalBooks = (int)countryTotal.Item2,
                        PercentageBooks = (float)countryTotal.Item3,
                        TotalPages = (int)countryTotal.Item4,
                        PercentagePages = (float)countryTotal.Item5
                    };
            }
        }
    }
}

File: LocalWebApp/ClientApp/src/app/Models/delta-books.ts
import { TallyDelta } from './tally-delta'

export interface ICategoryTotal {
    name: string;
    totalPages: number;
    totalBooks: number;
    percentagePages: number;
    percentageBooks: number;
};

export class CategoryTotal implements ICategoryTotal {
    constructor(
        public name: string = '',
        public totalPages: number = 0,
        public totalBooks: number = 0,
        public percentagePages: number = 0,
        public percentageBooks: number = 0) {
    }
}

export interface IDeltaBooks {
    date: Date;
    startDate: Date;
    daysSinceStart: number;
    overallTally: TallyDelta;
    lastTenTally: TallyDelta;
    languageTotals: ICategoryTotal[];
    countryTotals: ICategoryTotal[];
};

export class DeltaBooks implements IDeltaBooks {
    constructor(
        public date: Date = new Date(),
        public startDate: Date = new Date(),
        public daysSinceStart: number = 1,
        public overallTally: TallyDelta = new TallyDelta(),
        public lastTenTally: TallyDelta = new TallyDelta(),
        public languageTotals: ICategoryTotal[] = new Array<CategoryTotal>(),
        public countryTotals: ICategoryTotal[] = new Array<CategoryTotal>()) {
    }
}
